spec/models/repository/git_spec.rb

```diff
@@ -0,0 +1,363 @@
+          changeset = instance.find_changeset_by_name(r1)
+          expect(changeset.previous).to be_nil
+        end
+      end
+
+      it 'should find next changeset' do
+        %w|64f1f3e89ad1cb57976ff0ad99a107012ba3481d 64f1f3e89ad1|.each do |r2|
+          changeset = instance.find_changeset_by_name(r2)
+          %w|1ca7f5ed374f3cb31a93ae5215c2e25cc6ec5127 1ca7f5ed|.each do |r1|
+            expect(instance.find_changeset_by_name(r1)).to eq(changeset.next)
+          end
+        end
+      end
+
+      it 'should next nil' do
+        %w|1ca7f5ed374f3cb31a93ae5215c2e25cc6ec5127 1ca7f5ed|.each do |r1|
+          changeset = instance.find_changeset_by_name(r1)
+          expect(changeset.next).to be_nil
+        end
+      end
+
+      context 'with an admin browsing activity' do
+        let(:user) { FactoryGirl.create(:admin) }
+        let(:project) { FactoryGirl.create(:project) }
+
+        def find_events(user, options = {})
+          fetcher = Redmine::Activity::Fetcher.new(user, options)
+          fetcher.scope = ['changesets']
+          fetcher.events(Date.today - 30, Date.today + 1)
+        end
+
+        it 'should activities' do
+          c = Changeset.create(repository: instance,
+                               committed_on: Time.now,
+                               revision: 'abc7234cb2750b63f47bff735edc50a1c0a433c2',
+                               scmid:    'abc7234cb2750b63f47bff735edc50a1c0a433c2',
+                               comments: 'test')
+
+          event = find_events(user).first
+          assert event.event_title.include?('abc7234c:')
+          assert event.event_path =~ /\?rev=abc7234cb2750b63f47bff735edc50a1c0a433c2$/
+        end
+      end
+
+      describe 'encoding' do
+        let(:felix_hex) { "Felix Sch\xC3\xA4fer" }
+
+        it 'should display UTF-8' do
+          c = instance.changesets.where(revision: 'ed5bb786bbda2dee66a2d50faf51429dbc043a7b').first
+          expect(c.committer).to eq("#{felix_hex} <felix@fachschaften.org>")
+          expect(c.committer).to eq('Felix Schäfer <felix@fachschaften.org>')
+        end
+
+        context 'with latin-1 encoding' do
+          let (:encoding) { 'ISO-8859-1' }
+          let (:char1_hex) { "\xc3\x9c".force_encoding('UTF-8') }
+
+          it 'should latest changesets latin 1 dir' do
+            instance.fetch_changesets
+            instance.reload
+            changesets = instance.latest_changesets(
+              "latin-1-dir/test-#{char1_hex}-subdir", '1ca7f5ed')
+            expect(changesets.map(&:revision))
+              .to eq(['1ca7f5ed374f3cb31a93ae5215c2e25cc6ec5127'])
+          end
+
+          it 'should browse changesets' do
+            changesets = instance.latest_changesets(
+              "latin-1-dir/test-#{char1_hex}-2.txt", '64f1f3e89')
+            expect(changesets.map(&:revision))
+              .to eq([
+                    '64f1f3e89ad1cb57976ff0ad99a107012ba3481d',
```

Comment:
Use 2 spaces for indentation in an array, relative to the start of the line where the left bracket is.